### ISSUE TITLE
Barriers for speech recognition users

### ISSUE THREAD

**Issue by jessemenn** [wagtail/wagtail#9666]
### Issue Summary

These were raised during a demo from a Dragon AT user. Hopefully I've represented them correctly. 

Wagtail has some barriers for Dragon AT, and likely other speech recognition software, users.

Dragon crashes Wagtail 3's rich text editor. In Wagtail 4, Dragon does not recognize the rich text editor as an editable field.

Wagtail 4 UI has more issues:

- Unable to engage with minimap as minimap only engages via hover
- Breadcrumbs are collapsed until a hover interaction

Regarding links, buttons: Dragon users activate elements by saying 'click link' or 'click button' or 'click name' which prompts the software to attempt to click the name element, or generate an overlay of numbers associated with buttons or links, respectively. This works when those elements are visible, but many are hidden in Wagtail 4's new UI.

![Wagtail 4 content page click button](https://user-images.githubusercontent.com/1834167/201708150-4a0360a9-9aed-4e35-83c7-73611b6ed5b9.png)

- When editing a page, the up-chevron next to save draft (which exposes the publish and submit for moderation actions) is neither targetable nor usable 
- Previously, was able to dictate 'click collapse introduction' to collapse an introduction field of a page, but that now collapses all sections of a page

<img width="402" alt="save-draft-chevron" src="https://user-images.githubusercontent.com/1834167/201707360-581872ee-ca3b-4a4a-8100-8d55619d2d94.png">

Dictating 'click button' or 'click link' for the sidebar shows that some items are buttons (search, pages, reports, settings) while others are links (images, documents, snippets). This causes some frustration as it's not possible for all of the menu items to be labeled / available from a single command. In the attached screenshot for 'click link', unsure why two 'Live' are not labeled.

![Wagtail 4 home page click link](https://user-images.githubusercontent.com/1834167/201706835-270e5963-df39-4ac7-9cca-ade7a10054df.png)

![Wagtail 4 home page click button](https://user-images.githubusercontent.com/1834167/201707063-829f5dad-6984-4896-ba1c-670935cfbb34.png)

### Steps to Reproduce

1. Install Wagtail 3 or 4
2. Use the admin area with Dragon

- I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: no

### Technical details

- Wagtail 3; 4x
- Unsure of Dragon version, presumably all

**Comment by jessemenn** [comment]
User alerts me that I misstated:

> Previously, was able to dictate 'click collapse introduction' to collapse an introduction field of a page, but that now collapses all sections of a page

From user:

Using Wagtail 3, I've never actually said "click collapse…" for showing or hiding a field. It's only in Wagtail 4, where I experimented with the new expandable and collapsible sections, where I have tried to use voice to expand or collapse those sections. It's because the chevron behaviour suggests the words "expand" and "collapse" that I assumed saying "click collapse" or "click expand" might activate them.

**Comment by thibaudcolas** [comment]
Just reviewing everything here in order of appearance – reiterating what @Stormheg has said quite a bit.

## Minimap reveal

> Unable to engage with minimap as minimap only engages via hover

Technically the minimap does have a toggle button for keyboard, screen, and touch screen users. But we’ve configured the toggle button to only be shown on devices that _aren’t_ capable of hover interactions – which I’m guessing is the issue here: the (desktop) browser reports hover support, but that’s not actually possible with speech recognition.

So – we’re going to revisit this so the button is always present (#10055).

## Breadcrumbs reveal

> Breadcrumbs are collapsed until a hover interaction

Similarly here, this is true, but breadcrumbs also have a toggle button for keyboard / touch screen, and I’d assume also speech recognition users. We’re going to revisit the pattern in #9498 but I’m unclear whether there is any change needed specifically for speech recognition users.

## Footer actions

> When editing a page, the up-chevron next to save draft (which exposes the publish and submit for moderation actions) is neither targetable nor usable

This is opened at #7366, which we currently hope to tackle via #9030.

## Section toggles

> Previously, was able to dictate 'click collapse introduction' to collapse an introduction field of a page, but that now collapses all sections of a page

This might just be a matter of label? The accessible name of those buttons is "Toggle section" in both states, with an accessible description matching the section’s heading (so "Introduction" for a top-level field).

Would introducing tooltips for icon-only buttons like this help? For the label, any guidance on how to name those buttons well for speech recognition users would be very welcome.

## Sidebar navigation

> Dictating 'click button' or 'click link' for the sidebar shows that some items are buttons (search, pages, reports, settings) while others are links (images, documents, snippets). This causes some frustration as it's not possible for all of the menu items to be labeled / available from a single command. In the attached screenshot for 'click link', unsure why two 'Live' are not labeled.

As far as I understand this is intended. Menu items that reveal sub-menus are coded as buttons, menu items that directly go to a page are links. If there is a better pattern for us to follow here please advise.

Re the "Live" not being numbered – what @Stormheg said about this looking the same despite it not being a link.

---

As a general way to make progress on this I think we need:

- Guidance on how to name icon-only controls in a way that makes sense to speech recognition users. At the moment the naming tends to be decided by developers at the point of implementing the designs, so consistency is poor.
  - A particular point we’re inconsistent on is whether the label should convey the state of the control or not ("Toggle" vs. "Collapse/Expand").
  - And whether / how much we should name things according to the icon vs. what the feature is otherwise called
- Guidance on how to make the names of icon-only buttons more discoverable.
  - Widespread usage of tooltips?
  - Or avoid icon-only buttons as much as possible?
- List of the most problematic patterns on here

**Comment by thibaudcolas** [comment]
Just a note that #10286 has fixed what I believe is the most problematic of all those issues. #7366 is also very problematic but has been there for years (…). Either @laymonage or I will be working on that issue shortly, skipping the new designs we have of this component so we can patch this issue in existing releases (4.1, 4.2).

**Comment by thibaudcolas** [comment]
FYI the minimap reveal has now been changed to click only in #10344.

**Comment by thibaudcolas** [comment]
Revisiting the list above,

## Minimap reveal

> Unable to engage with minimap as minimap only engages via hover

This is fixed (#10055 / #10344)

## Breadcrumbs reveal

> Breadcrumbs are collapsed until a hover interaction

Still pending in #9498, though again I don’t think this is a blocker (breadcrumbs have a clickable toggle too)

## Footer actions

> When editing a page, the up-chevron next to save draft (which exposes the publish and submit for moderation actions) is neither targetable nor usable

See #7366 / #9030. Scheduled to be fixed in time for the August 2023 release.

## Section toggles

> Previously, was able to dictate 'click collapse introduction' to collapse an introduction field of a page, but that now collapses all sections of a page

TBC based on above comments.

## Sidebar navigation

> Dictating 'click button' or 'click link' for the sidebar shows that some items are buttons (search, pages, reports, settings) while others are links (images, documents, snippets). This causes some frustration as it's not possible for all of the menu items to be labeled / available from a single command. In the attached screenshot for 'click link', unsure why two 'Live' are not labeled.

TBC based on above comments.

**Comment by albfreeman** [comment]
### Wagtail userbar issue
The Wagtail userbar that appears at the bottom of live and draft pages is no longer usable with Dragon. I could previously use Dragon to activate this button in Wagtail 3. But it doesn't work in Wagtail 5 (I didn't try it when I tested Wagtail 4).

When I was viewing a live (or draft) page in Wagtail 3 I could say "click Wagtail" and this would bring up the userbar pop up menu. I could also say "click button" and then choose the number that corresponded to the Wagtail icon, just as I would do with anything else coded as a button. It doesn't respond to either of these commands in Wagtail 5.

If I use my mouse to activate the userbar, I can then use Dragon to select one of the options in the pop-up menu. For example, I can say "click edit this page" to open the visible page in the CMS. It's just the userbar itself that is no longer recognised by Dragon.

I'm using Dragon 15 and Firefox is my browser.

**Comment by laymonage** [comment]
@albfreeman We did some digging into this and apparently the correct label for the userbar is "View wagtail quick actions", see: https://github.com/wagtail/wagtail/issues/10674#issuecomment-1634016619

Could you confirm if commanding "Click view Wagtail quick actions" works with Dragon?

**Comment by thibaudcolas** [comment]
I believe all the issues here have been fixed? except #10674. For other items, there were other separate issues already created. Follow-up of sorts: #13429
